### PR TITLE
Make sure prepostfix boundaries are respected.

### LINKED ISSUE #189: Prepostfix boundary does not get properly cloned; missing a test?
Just discovered this in `OpScope.pm`:

    has $!prepostfix-boundary = 0;

    # meanwhile, later in the file:
    method clone {
        my $opl = self.new(
            infixprec => @.infixprec.map(*.clone),
            prepostfixprec => @.prepostfixprec.map(*.clone),
            :$!prepostfix-boundary,    # <--- *this* line
        );
        # ...stuff ...
    }

The intent is clear: we want to clone the `$!prepostfix-boundary` attribute with the rest of the opscope. Of course.

But private attributes aren't settable through the default `new`, as demonstrated here:

    $ perl6 -e 'class C { has $!priv; method set($n) { $!priv = $n }; method divulge { $!priv }; method clone { self.new(:$!priv) } }; given C.new { .set(2); say .divulge; my $c2 = .clone; say $c2.divulge }'
    2
    (Any)

So

* The indicated line does nothing, and
* very likely this would fail a test we never wrote.

What test? Basically a recapitulation of the above demonstration, but in the context of the boundary between prefix and postfix ops. (This, by the way, is something I should write down somewhere. I see [operators.md](https://github.com/masak/007/blob/69b7ceb607055cf23c93f13c999ae5edec27f840/operators.md) oddly has no mention of it. The commit that adds the attribute (and the bug), 60937955dbc93db6ad078e1822b7069fc4551a0a, doesn't mention the boundary at all.)

Ok, let's see.

    sub prefix:<&>(x) {                         #  op order: &
        return x ~ " prefix:<&>";
    }

    sub postfix:<!>(x) is looser(prefix:<&>) {  #  op order: ! & (tightest)
        return x ~ " postfix:<!>";
    }

    {
        sub prefix:<$>(x) {                     #  op order: $ ! & (tightest)
            return x ~ " prefix:<$>";           # should be: ! & $ (tightest)
        }

        say($&"application order:"!);           #   outputs: application order: prefix:<&> postfix:<!> prefix:<$>
                                                # should be: application order: prefix:<&> prefix:<$> postfix:<!>
    }

Prefixes never compete for precedence on their own, because syntactically they get a natural nesting and inside-out evaluation order. But they can compete *via a postfix*, since prefixes and postfixes share a precedence space.

The statement block above is exactly so that an opscope cloning will happen and we will lose our prepostfix-boundary (aka "the bug"). Had we kept that boundary, `prefix:<$>` would have been installed to the right of (=tighter than) `prefix:<&>`. But now it gets installed on its left, and the print statement demonstrates that.

Phew! No wonder we didn't have a test for this!

In fact, we could exploit our tightest built-in prefix op for the same effect, with a bit less code:

    sub postfix:<!>(x) is looser(prefix:<^>) {  #  op order: ! ^ (tightest)
        return [];
    }

    sub prefix:<$>(x) {                         #  op order: $ ! ^ (tightest)
        return x.elems();                       # should be: ! ^ $ (tightest)
    }

    say($^5!);                                  #   outputs: 0
                                                # should be: []

...but this code is possibly coupled to the default built-ins, which might after all change in the future for good reasons. I think I like the longer test better.

Ok, that was a lot of talking. The task here is simple: add the test. See it fail. Make `$!prepostfix-boundary` public, so that it doesn't get clobbered during a clone. See the test pass. So it's a test and a one-line fix.

### PR DESCRIPTION
**Pull request by vendethiel**
Fixes #189

### CODE REVIEW

File: t/features/custom-ops.t [hunk 1]
```diff
@@ -504,4 +504,20 @@ use _007::Test;
     outputs $program, "James\n", "can declare an operator with a backslash in the name";
 }
 
+{
+    my $program = q:to/./;
+        sub postfix:<!>(x) is looser(prefix:<^>) {
```
Comment (masak) [inline t/features/custom-ops.t]: Hm, may I suggest while #190 is open that we don't create operators `postfix:<!>` &mdash; they might make tests fail for the wrong reason.

**Comment by masak** [comment]
Looks good! Wrote a short line comment about a small concern.

Also, I think we might want to have that other, longer test in there. This shorter test is a little bit too tightly coupled with the builtins, which may change in the future. In fact if I had to choose between the two tests, I'd prefer the longer one. (But both are fine, too.)

**Comment by vendethiel** [comment]
Uhhhh.

```
not ok 33 - non-associativity inherits through the 'is equal' trait

# Failed test 'non-associativity inherits through the 'is equal' trait'
# at ~/007/lib/_007/Test.pm (_007::Test) line 324
```

**Comment by masak** [comment]
Huh, this patch has another failing test in `custom-ops.t`, according to Travis:

    not ok 33 - non-associativity inherits through the 'is equal' trait

Going to try and build locally to see if I can reproduce the failure.

**Comment by masak** [comment]
Isolating the test and running it in its own file makes it pass. Which means there's interference between one or more earlier tests and this one. Going to try to golf the failure.

**Comment by masak** [comment]
...actually, this might be related to #190, since the test declares a `postfix:<!>`.

**Comment by masak** [comment]
Yep. Commit coming up.

**Comment by masak** [comment]
Pushed as 7bef7bc4f4e9afff1bb7617ee4b4341bac2e8349 on master. Try rebasing on this and running the tests again.

**Comment by vendethiel** [comment]
Let's seee.

**Comment by masak** [comment]
Of course, it's a slight bit concerning that part of the issue is that the tests are not properly isolated from each other. That in itself is a possible separate avenue of investigation. I started down the path of golfing the test failure, but I abandoned it when I realized `postfix:<!>` was the culprit.

**Comment by vendethiel** [comment]
I'll wait to make sure this one passes now before adding the big brother test.

**Comment by masak** [comment]
Sure thing.

It should pass. I rebase and tested locally, and that passed.

**Comment by masak** [comment]
I'm willing to merge this now. Should I wait for the commit with the big test, or can we handle that after merge?

**Comment by vendethiel** [comment]
Right up ahead.

**Comment by masak** [comment]
Looks good. Will merge, right after some commits I'm about to push.

**Comment by masak** [comment]
Merged. Thanks!

**Pull request closed without merging**